### DISCUSSION TITLE
Plugin for Search

### DISCUSSION THREAD

**Discussion by theodorechu** [unifiedjs/unified#109] [Ideas] [answered]
I'm building an editor extension for [Standard Notes](https://standardnotes.org) called the [Append Editor](https://beta.appendeditor.com) ([Source](https://github.com/theodorechu/append-editor)). It makes heavy use of Remark and Rehype plugins to parse Markdown into React, so I'm very thankful for everything that the Remark and Rehype plugins already do. 

I would like to build a feature to search within the editor, which is sometimes embedded within an `iframe` and does not have access to a browser's built-in search. I'm a new developer, so I don't have much experience building search inside applications. 

I thought a potential solution would be to have a plugin for Remark, or probably Rehype, to allow for searching a keyword within the parsed/rendered html. For example, Remark and Rehype would parse the following Markdown text into the following html:

Markdown:
```md
# Title

We want to search for a keyword within this text.
```
HTML:
```html
<h1>Title</h1>
<p>We want to search for a keyword within this text.</p>
```

With this potential plugin, I am imagining that we would specify a keyword, such as `keyword`, and a class name, such as `searchMatch` and get something like this:

```html
<h1>Title</h1>
<p>We want to search for a <span class="searchMatch">keyword</span> within this text.</p>
```

Is something like this possible and/or feasible with Remark/Rehype plugins? Would the plugin be more feasible as a Remark or Rehype plugin? Or is this outside the scope of what Unified tries to accomplish?

Thanks for taking the time to consider this.

**Comment by ChristianMurphy** [comment, marked as answer]
@TheodoreChu can you expand on two things:
1. what do you mean by "search"? (exact text matches, partial text matches, similarity scoring and ranking, semantic scoring and ranking)
2. what would you be searching? (a section of a document, a full document, many documents)

Broadly when I hear search I tend to think of [Solr](https://lucene.apache.org/solr/), [ElastiSearch](https://www.elastic.co/), [Sphinx](http://sphinxsearch.com/), [MeiliSearch](https://docs.meilisearch.com/), etc. Full featured search engines, which have support for structured documents like Markdown or HTML, that have been processed with Unified.

Unified does have some single document search capabilities https://github.com/syntax-tree/nlcst-search for example. Which may work for some use-cases, but generally Unified is more focused on transforming a single document, than advanced cross document search.

**Reply by theodorechu** [reply]
Hi @ChristianMurphy, thanks for the response. I'm thinking about exact text matches in a single HTML or Markdown document. For example, suppose someone provides a single `.md` file or a string of Markdown text as well as another string of text that we call the keyword. The Remark and Rehype plugins transform the `.md` file or the string into a single HTML document. 

I'm wondering if it would be feasible for another Remark/Rehype plugin to find instances of the keyword in the `.md` file or Markdown text and wrap it in `<span>` tags with a specified `class` for custom styling while it's processing the Markdown into HTML. Suppose we stylize the class to have a different background (e.g., some shade of yellow). Then, the plugin would effectively work as a "search". If we compare an HTML document made without the extra plugin and an HTML document made with the plugin, the difference would be that the second has the specified keyword highlighted. The result in the latter document looks similar to what you would get if you had  webpage displaying the HTML of the former document and searched the webpage with ctrl/cmd+F in a browser. I guess I'm wondering if we can imitate ctrl/cmd+F in a browser by using Unified/Remark/Rehype plugins. Does that make more sense?

Also, thanks for the links. I am not familiar with most of them.

**Reply by ChristianMurphy** [reply]
It possible, here is a naive implementation (see 1. below) that could be used as a starting point https://astexplorer.net/#/gist/ed8574961c9ea4e5d6f772747b9a70da/03209e32513d1a4d05b48359d48f879e06e5e16d and you can find more about working with ASTs and plugins here: https://unifiedjs.com/learn/

There are some trade offs though:
1. handling search across constructs in an AST can be complex, for example if you want to match "hello world", but the text could be in links or emphasis
    ```markdown
    [hello][] [world][]
    *hello* *world*
    ```
    to do this would require some potentially complex logic to scan across nodes and determine where the match is and where the span could safely be added.
2. Every search (and if it is live, every character of every search) would trigger a fresh transform of the document, which can lock up the UI. This can be reduced some with [debouncing](https://davidwalsh.name/javascript-debounce-function) the search

**Reply by theodorechu** [reply]
Great, thank you @ChristianMurphy for the response and the suggestions. 😃

**Reply by wooorm** [reply]
https://github.com/syntax-tree/hast-util-find-and-replace might be useful for searching with a regex and wrapping it in a span btw! (but it has some limitation, which Christian points out above)

**Reply by Nickersoft** [reply]
@wooorm I just stumbled across this thread after searching for a similar solution for my project – in particular, needing to match across HTML elements (as @ChristianMurphy mentioned in his first "trade off"). I'm curious how hard it would be to bring this functionality to `hast-util-find-and-replace`. I had been looking at [mark.js](https://markjs.io/), which [implements this logic](https://github.com/julmot/mark.js/blob/7f7e9820514e2268918c2259b58aec3bd5f437f6/src/lib/mark.js#L583-L611), but being new to Unified + not fully understanding the inner workings of mark.js, I'm a bit at a loss of how to bring it to Unified. I figured it's less expensive to load an HTML string into Unified than to load it into JSDOM just so I can use mark.js.

@TheodoreChu Were you ever able to come up with a solid solution to search?

**Reply by wooorm** [reply]
As I understand it, it combines all `text` nodes into a string, and saves information on positions in that string back to text nodes. It regexes the whole string, converts those places to original nodes, and marks each partial match. This can be done in hast as well.

**Reply by Nickersoft** [reply]
Thanks for such a quick response @wooorm! Yeah, after I posted my comment and dug a bit deeper I discovered that as well, which seems to be a reasonable approach. I think what I'm unsure of at this stage is converting between a DOM-based approach and the AST-based approach. For example, mark.js uses the built-in `splitText()` method for `Text` nodes, which I'm unsure how to reproduce in Unified. It also uses properties such as [Node.previousSibling](https://developer.mozilla.org/en-US/docs/Web/API/Node/previousSibling), which I'm not sure Unified has an alternative for.

Also, let me know if there's a more suitable place I should migrate this convo, such as a GitHub issue or separate discussion thread.

**Reply by wooorm** [reply]
The DOM is a complex model with lots of APIs that you need to know.
unist is objects, arrays, and strings. So you can use normal functions like `slice` to split a string or an array. To access a previous sibling in unist, you do `parent.children[theIndex - 1]`.
The things you want to do, are also done by `hast-util-find-and-replace`, I think you’ll find most of what you’re looking for in its code.

As for where to ask things, that depends on where you want this conversation to go! If you think this can be added to `find-and-replace`, it could be there. If you want to make a new thing, a new discussion is probably better.
